p2p/server.py

```diff
@@ -82,136 +85,187 @@ def __init__(self,
 
-        # Now we loop over all of the devices attempting to setup a port
-        # mapping from their external IP to the internal IP.
+        # Now we loop over all of the devices until we find one that we can use.
         for device in devices:
             try:
-                connection = device.WANIPConn1
+                device.WANIPConn1
             except AttributeError:
                 continue
+            return device
+        return None
 
-            # Detect our internal IP address (or abort if we can't determine
-            # the internal IP address
-            for iface in netifaces.interfaces():
-                for _, addr in netifaces.ifaddresses(iface):
-                    network = addr['addr'].rstrip('.', 1)
-                    if network in device.location:
-                        internal_ip = addr['addr']
-                        break
-            else:
-                self.logger.warn(
-                    "Unable to detect internal IP address in order to setup NAT portmap"
-                )
-                continue
-
-            connection.AddPortMapping(
-                NewRemoteHost=connection.GetExternalIPAddress()['NewExternalIPAddress'],
-                NewExternalPort=self.server_address.tcp_port,
-                NewProtocol='TCP',
-                NewInternalPort=self.server_address.tcp_port,
-                NewInternalClient=internal_ip,
-                NewEnabled='1',
-                NewPortMappingDescription='Created by Py-EVM',
-                NewLeaseDuration=lifetime,
-            )
-            self.logger.info("NAT port forwarding successfully setup")
-            break
-        else:
-            self.logger.warning('Unable to setup port forwarding for NAT')
-
-    async def _start(self) -> None:
-        self._server = await asyncio.start_server(
+    async def _start_tcp_listener(self) -> None:
+        # TODO: Support IPv6 addresses as well.
+        self._tcp_listener = await asyncio.start_server(
             self.receive_handshake,
-            host=self.address.ip,
-            port=self.address.tcp_port,
+            host='0.0.0.0',
+            port=self.port,
         )
 
+    async def _close_tcp_listener(self) -> None:
+        self._tcp_listener.close()
+        await self._tcp_listener.wait_closed()
+
+    async def _start_udp_listener(self, discovery: DiscoveryProtocol) -> None:
+        loop = asyncio.get_event_loop()
+        # TODO: Support IPv6 addresses as well.
+        self._udp_transport, _ = await loop.create_datagram_endpoint(
+            lambda: discovery,
+            local_addr=('0.0.0.0', self.port),
+            family=socket.AF_INET)
+
+    async def _close_udp_listener(self) -> None:
+        cast(asyncio.DatagramTransport, self._udp_transport).abort()
+
     async def _close(self) -> None:
-        self._server.close()
-        await self._server.wait_closed()
+        await self._close_tcp_listener()
+        await self._close_udp_listener()
```

Comment:
Could these two lines be run in parallel with [`asyncio.gather`](https://docs.python.org/3.5/library/asyncio-task.html#asyncio.gather)

Reply:
Yep, I don't see why not!